Convert to Oauth v2 Support

14rynx opened the pull request:
This PR updates to ESI Oauth v2 which is now mandatory.

There are also some smaller changes that were in the branch I was using before that ended up in here:
- whoami output is now snake_case
- Preston can automatically retry requests with backoff and will raise for http status

Draft PR since I still need to verify the Oauth v1 to Oauth v2 automatic migration and write exactly what to do to migrate.

14rynx committed:
Added internal retry mechanisms so request can be made on a higher level.

14rynx committed:
Move url building into its own function

14rynx committed:
Allow direct pasting of scopes from developers.eveonline.com

14rynx committed:
Added methods to authorize from refresh token including legacy support.

14rynx committed:
Fix using reserved id name as argument.

14rynx committed:
Converted to Oauth V2

14rynx committed:
Added refresh token update callback function.

14rynx committed:
Rename authenticate form refresh token function

14rynx commented:
Looks like the upgrade mechanism is no longer active, so I will remove that part of the code.

There are the following changes to preston in this PR. The basic authentication flow is the same but I had to tweak things here and there.

1. Since CCPs Oauth V2 requires an application state, you can now pass one to `preston.get_authorize_url()` otherwise the state `default` is used.

2. Since Oauth V2 tokens rotate, it is mandatory to save them once they get changed. To do that Preston now supports
a `refresh_token_callback` that can be passed at startup or set sometimes later. This should be a function that takes a Preston instance as argument and saves it's token somewhere e.g.
  ```python
  def my_refresh_token_callback(preston: Preston):
       character_id = preston.whoami().get("character_id")
       long_lasting_storage[character_id] = preston.refresh_token
  
  preston_instance = Preston(
    ...
    refresh_token_callback=my_refresh_token_callback
  )
   ```
   
 3. The return contents of the `whoami()` function changed. It now returns
  ```python
  {
    "character_id": ...,
    "character_name": ...,
    "scopes": ...,
    "owner_hash": ...,
    "token_type": ...,
  }
  ```

4. Preston now supports automatic retries of status codes with typical network / ESI problems (HTTP Statuses 500, 502, 503, 504, JsonDecodeErrors, Timeouts) with exponential backoff and respects ESI error limits (HTTP Statuses 420, 429). You can set the number of retries and the timeout in the instantiation. Preston will now raise for HTTP Status, so you can build propper error handling and don't have to just rely on the error body. For example:
  ```python
  try:
      authed_preston = preston.authenticate_from_refresh(character.token)
  except HTTPError as exp:
      # Some kind of error the user has to deal with e.g. no permissions, token revoked on website
      ...
  except ConnectionError as exp:
      # Network issue. If it is a background task probably fine to log and skip, if there is a user tell him network is bad
      ...
      # in both cases The exception contains the full response e.g. you can
      print(exp.response.status_code)
      print(exp.response.text)
      # etc.
  ```

5. You can now copy a preston instance but make it use another refresh token:
  ```python
    auth = preston.authenticate_from_refresh('their_refresh_token_here')
  ```
  (this was intended to enable old token upgrades as well, but unfortunately that is disabled already by CCP)

14rynx committed:
Remove legacy token upgrade mechanism and update legacy token warning.

Celeo reviewed (commented):
Hi, thanks so much for your work on this! I love the retry and backoff idea.

A quick thought: the change to `Cache.set` means that the expiration header is not longer defaulting to `0`, which means the tests in `tests/test_cache.py` fail; I don't know if this will impact the normal running of programs, or if the tests just need to be updated.

14rynx committed:
Update tests to new functionality

14rynx commented:
Good catch! I updated the cache.set at the same time as all calls to it so there shouldn't be a problem. I modified the tests to now test the new behavior.

Currently I am thinking that some of the changes could be good to have in the README.md, what do you think?

Celeo commented:
Thanks for updating!

Sure, mentioning the retries and back-off would definitely be good to put in the README! Probably somewhere in the "Usage" section.

14rynx committed:
Update README.md for oauth V2 and error handling.

14rynx committed:
Made refresh_token_callback optional.

14rynx commented:
I added some pointers in the README and made sure I don't break usecases where you don't store the refresh token externally.

Celeo committed:
Update README.md

Celeo commented:
Wonderful!
I made some minor README tweaks, but nothing crazy.

Celeo commented:
Thanks for the contribution!
